R/register_extendr.R

```diff
@@ -36,39 +36,54 @@ register_extendr <- function(path = ".", quiet = FALSE, force_wrappers = FALSE)
 
   outfile <- rprojroot::find_package_root_file("R", "extendr-wrappers.R", path = path)
 
-  # If force_wrappers is TRUE, use tryCatch() to generate minimal wrappers even
-  # when there's some error (e.g. the symbol cannot be found).
-  # If FALSE, execute make_wrappers() only when the package can be loaded.
+  # If force_wrappers is TRUE, generate minimal wrappers even when there's some
+  # error (e.g. the symbol cannot be found).
   if (isTRUE(force_wrappers)) {
-    tryCatch(
-      make_wrappers(pkg_name, pkg_name, outfile, use_symbols = TRUE, quiet = quiet),
-      error = function(...) {
-        warning(
-          "Generating the wrapper functions failed, so a minimal one is used instead",
-          call. = FALSE
-        )
-        make_example_wrappers(pkg_name, outfile)
-      }
-    )
-  } else if (requireNamespace(pkg_name, quietly = TRUE)) {
-    make_wrappers(pkg_name, pkg_name, outfile, use_symbols = TRUE, quiet = quiet)
+    error_handle <- function(e) {
+      msg <- "Generating the wrapper functions failed, so a minimal one is used instead"
+      warning(msg, call. = FALSE)
+      make_example_wrappers(pkg_name, outfile)
+    }
   } else {
-    stop(
-      glue("Package {pkg_name} cannot be loaded. No wrapper functions were generated."),
-      call. = FALSE
-    )
+    error_handle <- function(e) {
+      stop("Generating the wrapper functions failed", call. = FALSE)
+    }
   }
+
+  tryCatch(
+    make_wrappers(pkg_name, pkg_name, outfile,
+      use_symbols = TRUE, quiet = quiet,
+      # Call the wrapper generation in a separate R process to avoid the problem
+      # of loading and unloading the same name of library (c.f. #64).
+      use_callr = TRUE
+    ),
+    error = error_handle
+  )
 }
 
 make_wrappers <- function(module_name, package_name, outfile,
-                          use_symbols = FALSE, quiet = FALSE) {
+                          use_symbols = FALSE, quiet = FALSE,
+                          use_callr = FALSE) {
   wrapper_function <- glue("wrap__make_{module_name}_wrappers")
-  x <- .Call(
+
+  func <- function(package_root, ...) {
+    pkgload::load_all(package_root, quiet = TRUE)
```

Comment:
I don't think this line should be executed when we're running `rust_source()`. I'm still wondering whether it makes sense to try to let `make_wrappers()` handle two so very different use cases. It feels forced.

Reply:
Ah, true...